src/Microsoft.AspNet.Mvc.TagHelpers/LabelTagHelper.cs

```diff
@@ -45,12 +45,21 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
                 {
                     output.MergeAttributes(tagBuilder);
 
+                    var childContent = await context.GetChildContentAsync();
+
                     // We check for whitespace to detect scenarios such as:
                     // <label for="Name">
                     // </label>
-                    if (string.IsNullOrWhiteSpace(output.Content))
+                    if (string.IsNullOrWhiteSpace(childContent))
+                    {
+                        if (string.IsNullOrEmpty(output.Content))
+                        {
+                            output.Content = tagBuilder.InnerHtml;
+                        }
+                    }
+                    else if (string.IsNullOrEmpty(output.Content))
                     {
-                        output.Content = tagBuilder.InnerHtml;
+                        output.Content = childContent;
```

Comment:
nit: this is hard on the eyes due to repeated conditions.  suggest instead
``` c#
  if (string.IsNullOrEmpty(output.Content))
  {
    if (string.IsNullOrWhiteSpace(childContent))
    {
      // Provide default label text since there was nothing useful in the Razor source.
      output.Content = tagBuilder.InnerHtml;
    }
    else
    {
      output.Content = childContent;
    }
  }
```